scylla/src/transport/load_balancing/latency_aware.rs

```diff
@@ -0,0 +1,295 @@
+                if let Some(nodes) = nodes {
+                    if nodes.is_empty() {
+                        error!("Empty node list - driver/policy bug!");
+                        return;
+                    }
+                    trace!("LatencyAwarePolicy: updating min average");
+
+                    let min_avg = nodes
+                        .iter()
+                        .filter_map(|node| {
+                            node.average_latency
+                                .read()
+                                .unwrap()
+                                .map(|timestamped_average| timestamped_average.average)
+                        })
+                        .min();
+                    if let Some(min_avg) = min_avg {
+                        min_latency.store(min_avg);
+                    }
+                }
+            }
+        }
+        .remote_handle();
+        tokio::task::spawn(updater_fut);
+
+        Self {
+            exclusion_threshold,
+            retry_period,
+            update_rate,
+            minimum_measurements,
+            last_min_latency: min_latency_clone,
+            nodes_tx,
+            ask_for_nodes: ask_for_nodes_clone,
+            _updater_handle: updater_handle,
+        }
+    }
+
+    fn refresh_last_min_avg_nodes(&self, nodes: &[Arc<Node>]) {
+        if self.ask_for_nodes.swap(false, Ordering::Relaxed) {
+            let _ = self.nodes_tx.update(Some(nodes.to_owned()));
+            trace!("LatencyAwarePolicy: updating nodes list for the min average updater");
+        }
+    }
+
+    fn make_plan<'a>(&self, plan: Vec<Arc<Node>>) -> Plan<'a> {
+        self.refresh_last_min_avg_nodes(&plan);
+        let min_avg_latency = match self.last_min_latency.load() {
+            Some(min_avg) => min_avg,
+            None => return Box::new(plan.into_iter()), // noop, as no latency data has been collected yet
+        };
+
+        Box::new(IteratorWithSkippedNodes::new(
+            Box::new(plan.into_iter()),
+            self.exclusion_threshold,
+            self.retry_period,
+            self.minimum_measurements,
+            min_avg_latency,
+        ))
+    }
+}
+
+impl Default for LatencyAwarePolicy {
+    fn default() -> Self {
+        Self::new(
+            2_f64,
+            Duration::from_secs(10),
+            Duration::from_millis(100),
+            50,
+        )
+    }
+}
+
+impl LoadBalancingPolicy for LatencyAwarePolicy {
+    fn plan<'a>(&self, _statement: &Statement, cluster: &'a ClusterData) -> Plan<'a> {
+        self.make_plan(cluster.all_nodes.clone())
+    }
+
+    fn name(&self) -> String {
+        "LatencyAwarePolicy".to_string()
+    }
+
+    fn requires_latency_measurements(&self) -> bool {
+        true
+    }
+}
+
+impl ChildLoadBalancingPolicy for LatencyAwarePolicy {
+    fn apply_child_policy(
+        &self,
+        plan: Vec<Arc<Node>>,
+    ) -> Box<dyn Iterator<Item = Arc<Node>> + Send + Sync> {
+        self.make_plan(plan)
```

Comment:
That's unfortunately true.
However, let's discuss what problems might arise from that. The direct consequence is that `min_avg` can be different that it should (more specifically, it can only be **greater** that it should be, due to the properties of `min` function called on a subset instead of the whole set). If `min_avg` is greater than it should be, some nodes may not be penalised while they should, and this can alter the desired order of querying. An example:
Node | avg latency (ms)
A       |  10
B       |  100
C       |  150
Let's assume that A was excluded due to token awareness, and so `min_avg` updater computed `100 ms` as the min_avg. If we now make a query that in aspect of token awareness is present on A and B, and moreover B is before A on the initial nodes list, then B won't be penalised and it will become the coordinator of the query, even though A is a much better candidate for that due to 10x smaller latency.